KeenClient/KeenClient.m

```diff
@@ -291,78 +228,140 @@ + (BOOL)validateKey:(NSString *)key {
+             andNetwork:(KIONetwork*)network
+               andStore:(KIODBStore*)store
+            andUploader:(KIOUploader*)uploader {
+    // Validate key parameters
     if (![KeenClient validateProjectID:projectID]) {
+        KCLog(@"Invalid projectID: %@", projectID);
         return nil;
     }
 
-    if (!dbStore) {
-        dbStore = [[KIODBStore alloc] init];
+    if (nil != writeKey) {
+        // only validate a non-nil value
+        KCLog(@"Invalid writeKey: %@", writeKey);
+        if (![KeenClient validateKey:writeKey]) {
+            return nil;
+        }
     }
 
-    self = [self init];
-    if (self) {
-        self.projectID = projectID;
-        if (writeKey) {
-            if (![KeenClient validateKey:writeKey]) {
-                return nil;
-            }
-            self.writeKey = writeKey;
-        }
-        if (readKey) {
-            if (![KeenClient validateKey:readKey]) {
-                return nil;
-            }
-            self.readKey = readKey;
+    if (nil != readKey) {
+        // only validate a non-nil value
+        KCLog(@"Invalid readKey: %@", readKey);
+        if (![KeenClient validateKey:readKey]) {
+            return nil;
         }
     }
 
+    self = [self initWithNetwork:network
+                        andStore:store
+                     andUploader:uploader];
+    if (nil != self) {
+        self.projectID = projectID;
+        self.writeKey = writeKey;
+        self.readKey = readKey;
+        self.uploader = uploader;
+    }
+
     return self;
 }
 
+- (instancetype)init {
+    return [self initWithNetwork:KIONetwork.sharedInstance
+                        andStore:KIODBStore.sharedInstance
+                     andUploader:KIOUploader.sharedInstance];
+}
+
+- (id)initWithProjectID:(NSString*)projectID
+            andWriteKey:(NSString*)writeKey
+             andReadKey:(NSString*)readKey {
+    return [self initWithProjectID:projectID
+                       andWriteKey:writeKey
+                        andReadKey:readKey
+                        andNetwork:KIONetwork.sharedInstance
+                          andStore:KIODBStore.sharedInstance
+                       andUploader:KIOUploader.sharedInstance];
+}
+
+
 # pragma mark - Get a shared client
 
-+ (KeenClient *)sharedClientWithProjectID:(NSString *)projectID andWriteKey:(NSString *)writeKey andReadKey:(NSString *)readKey {
-    if (!sharedClient) {
-        sharedClient = [[KeenClient alloc] init];
-    }
++ (KeenClient *)sharedClientWithProjectID:(NSString *)projectID
+                              andWriteKey:(NSString *)writeKey
+                               andReadKey:(NSString *)readKey {
+
+    // Validate key parameters
     if (![KeenClient validateProjectID:projectID]) {
+        KCLog(@"Invalid projectID: %@", projectID);
         return nil;
     }
 
-    if (!dbStore) {
-        dbStore = [[KIODBStore alloc] init];
-    }
-    sharedClient.projectID = projectID;
-
     if (writeKey) {
```

Comment:
Nope, I'll change it to nil != writeKey, which I generally prefer since there isn't a type coercion and the intent is clearer.